pkg/buildplan/filters.go

```diff
@@ -59,9 +59,20 @@ func FilterStateArtifacts() FilterArtifact {
 
 func FilterSuccessfulArtifacts() FilterArtifact {
 	return func(a *Artifact) bool {
-		return a.Status == types.ArtifactSucceeded ||
-			a.Status == types.ArtifactBlocked ||
-			a.Status == types.ArtifactStarted ||
-			a.Status == types.ArtifactReady
+		return a.Status == types.ArtifactSucceeded
+	}
+}
+
+func FilterFailedArtifacts() FilterArtifact {
+	return func(a *Artifact) bool {
+		return a.Status == types.ArtifactBlocked ||
+			a.Status == types.ArtifactFailedTransiently ||
+			a.Status == types.ArtifactFailedPermanently
+	}
+}
+
+func FilterNeedsBuild() FilterArtifact {
+	return func(a *Artifact) bool {
+		return a.Status != types.ArtifactSucceeded
```

Comment:
Is this true? A failed or skipped artifact does not need to be built and it won't be in the future.

Reply:
I've renamed it to `FilterNotBuild()`, because the code using this does in fact want artifacts that failed.